docs/decisions/0033-rag-in-sk.md

```diff
@@ -0,0 +1,297 @@
+- Custom user plugin.
+
+ChatGPT Retrieval Plugin:
+
+```csharp
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+// Import ChatGPT Retrieval Plugin using OpenAPI specification
+// https://github.com/openai/chatgpt-retrieval-plugin/blob/main/.well-known/openapi.yaml
+await kernel.ImportPluginFromOpenApiAsync("ChatGPTRetrievalPlugin", openApi!, executionParameters: new(authCallback: async (request, cancellationToken) =>
+{
+    request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", "chat-gpt-retrieval-plugin-token");
+}));
+
+const string Query = "What is my budget for 2024?";
+const string Prompt = "{{ChatGPTRetrievalPlugin.query_query_post queries=$queries}} {{$query}}";
+
+var arguments = new KernelArguments
+{
+    ["query"] = Query,
+    ["queries"] = JsonSerializer.Serialize(new List<object> { new { query = Query, top_k = 1 } }),
+};
+
+var result = await kernel.InvokePromptAsync(Prompt, arguments);
+```
+
+TextMemoryPlugin:
+
+```csharp
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+// NOTE: If the decision will be to continue support memory-related public API, then it should be revisited.
+// It should be up-to-date with new Semantic Kernel patterns.
+// Example: instead of `WithChromaMemoryStore`, it should be `AddChromaMemoryStore`.
+var memory = new MemoryBuilder()
+    .WithChromaMemoryStore("https://chroma-endpoint")
+    .WithOpenAITextEmbeddingGeneration("text-embedding-ada-002", "api-key")
+    .Build();
+
+kernel.ImportPluginFromObject(new TextMemoryPlugin(memory));
+
+var result = await kernel.InvokePromptAsync("{{recall 'Company budget by year'}} What is my budget for 2024?");
+```
+
+Custom user plugin:
+
+```csharp
+public class MyDataPlugin
+{
+    [KernelFunction("search")]
+    public async Task<string> SearchAsync(string query)
+    {
+        // Make a call to vector DB and return results.
+        // Here developer can use already existing .NET SDK from specific vector DB provider.
+        // It's also possible to re-use Semantic Kernel memory connector directly here: 
+        // new ChromaMemoryStore(...).GetNearestMatchAsync(...)
+    }
+}
+
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+kernel.ImportPluginFromType<MyDataPlugin>();
+
+var result = await kernel.InvokePromptAsync("{{search 'Company budget by year'}} What is my budget for 2024?");
+```
+
+The reason why custom user plugin is more flexible than `TextMemoryPlugin` is because `TextMemoryPlugin` requires all vector DBs to implement `IMemoryStore` interface with disadvantages described above, while custom user plugin can be implemented in a way of developer's choice. There won't be any restrictions on DB record schema or requirement to implement specific interface.
```

Comment:
Is there a way to improve `TextMemoryPlugin` so that it doesn't have as many of the issues? I agree that a "custom" plugin is the best escape valve (that's what most/all of our customers do to get hybrid search support), but hopefully we can make this a more rare occurrence.

Reply:
There is always space for improvement :) 
I think the main disadvantage of `TextMemoryPlugin` is the requirement for all integrations to implement the same interface and have the same data schema, which is not very flexible. I think it works well for demonstration/experimentation purposes, but for more flexibility and easier memory-related extension in the future, personally I would probably go with custom plugin path, configure and use it in a way that covers my scenario the best.

In short:
- TextMemoryPlugin - easier setup, less flexibility
- Custom plugin - harder setup, more flexibility

Reply:
I agree with the general analysis here, i.e. that IMemoryStore imposes certain limitations that may lead some users to switch to direct interaction with their vector DB, foregoing the abstraction. However, it would be good to understand how prevalent that actually is... In many cases, requirements from metadata stored in the vector db simply aren't that advanced/specific, and the somewhat simplistic IMemoryStore abstraction isn't an actual hindrance.

Otherwise, I think it's not a bad story to have the abstraction as a way to get up and running really quickly - without having to mess around with a custom plugin and understanding the specific vector DB's SDK; then, as new requirements come in, some users may "graduate" out of that story to interact directly with the vector DB, if they really need to.

Reply:
That makes sense! I think if we add more advanced examples how to quickly use memory with our abstractions and how to interact directly with the vector DB for better flexibility, we should cover most of the cases for users.

Reply:
To expand over the points made and point 5 in the out-of-scope section, My app (which uses Neo4j) required the modification of the vector retrieval query at runtime. I ended up creating the [neo4j connector](https://www.nuget.org/packages/Codeblaze.SemanticKernel.Connectors.Memory.Neo4j), which takes in a query factory that can be overridden, giving control to the user over db operations done by the semantic kernel.

The biggest thing that felt odd to me was that I only required a `Read Only Memory Connector` since we have a separate dashboard and API for writing to the database and don't want our AI-enabled APIs to have any write capabilities over the database. Extending such a read-only connector would allow integration with traditional search API.

Consider a use case where I am building a discord bot that references existing messages sent by other users on the channel. In such a case, I don't require embedding search as I can use the existing discord search API and retrieve the required messages.

Reply:
I think a good litmus test is that a developer should (at the very least) be able to take an index they've built for LangChain and/or the ChatGPT Retrieval Plugin and be able to quickly do "simple" queries using our connectors. I think this test would encourage the following principels:
1. Devs should be able to bring their own schema
2. We should try to avoid requiring "non-standard" fields (e.g., IsReferenceField) that make import difficult/impossible (if someone needs "advanced" features like this, they should build a custom connector)
3. "Easy" optional configuration should still be possible on a connector-by-connector basis (for example, we should have config for Azure AI Search that has similar behavior as the [ChatGPT Retrieval plugin env variables for Azure AI Search](https://github.com/openai/chatgpt-retrieval-plugin/blob/main/docs/providers/azuresearch/setup.md))

Reply:
> The biggest thing that felt odd to me was that I only required a Read Only Memory Connector since we have a separate dashboard and API for writing to the database and don't want our AI-enabled APIs to have any write capabilities over the database.

FWIW this is a very odd use-case for me... if you need actual access control (allow read but not write), then that's a feature you need from the database itself; just hiding the write APIs at the SDK level does very little, and can be easily circumvented if someone wants to do that. A read-only SDK/driver is only to give the illusion of providing access control without actually doing it, which is in fact more dangerous.

On the other hand, someone may want a read-only SDK not for security purposes, but just to prevent **accidental** invocation of write APIs, as a very defensive way to guard against programmer error. That's very easy to do without writing dropping the abstraction and writing an SDK/connector - you can simply have a trivial ReadOnlyMemoryStore which wraps an existing Semantic Kernel IMemoryStore, delegates the read APIs to the wrapped memory store and throws from the write APIs. If that's deemed to be actually useful to a lot of people (not too sure about that), that's something that can be added to SK itself.

So I think the read-only access is a rather weak case for foregoing the abstraction.

Reply:
> "Easy" optional configuration should still be possible on a connector-by-connector basis

Remember that the abstraction covers using the vector DB (reading/writing), but you're still the one configuring and instantiating the IMemoryStore implementation. In other words, you should be able to do up-front configuration of your specific vector DB in whichever way you want, and then the resulting IMemoryStore can be used with the standard SK APIs etc. So here again, I don't see where the abstraction limits configurability.

Reply:
> So I think the read-only access is a rather weak case for foregoing the abstraction.

@roji We discussed it with SK team, and it looks like having an abstraction for read-only operations (like get a record by key or get a record using vector search) should be enough at this point, because the main idea behind memory connectors is the ability to import it as plugin and then re-use it directly in prompts for RAG scenarios, using following prompt as example:

Prompt: `{{search 'company budget by year'}} What is my budget for 2024?` 
LLM Response: `Your budget for 2024 is ...`

As for write operations, we don't expect users to save data or perform any kind of data manipulation or DB architecture changes (like creating collections/indexes) from prompt. For data operations like `add`, `update`, `remove` - users should probably handle it separately from Kernel.

So, we are thinking about leaving current `IMemoryStore` as it is, with read and write methods for experimentation purposes and quick setup, while providing new lightweight abstraction for search operations, that will support custom schema for more flexibility, so it will allow to use RAG pattern and perform vector search for already existing indexes/collections in user databases.

What do you think about it?

Reply:
I think I'd try to separate the two problems, i.e. the custom schema question, and the read/write vs. read-only question... I'd recommend trying to avoid introducing a **new** abstraction, since that would require reimplementing connectors for all the vector databases, and then maintaining the two implementations for each one etc.

Instead, if the goal is to support flexible schemas, I'd rather see about how IMemoryStore can be evolved to support that (IIRC the abstraction is still considered experimental and therefore can be changed in breaking ways, right?). This might indeed be tricky given that you have quite a few implementations which you ideally want to keep working, but I'd at least give this a solid try.

Re read-only, I agree that users are unlikely to need to do data updates from prompt. However, as I wrote above I'm not sure I see a specific problem with using an abstraction (IMemoryStore) in a read-only context (prompt), where that abstraction happens to also support write operations; there's again no actual security aspect here (since just not having the write APIs in the abstraction doesn't provide security, only hiding), so the existence of the write APIs shouldn't be a problem in itself.

In addition, even if you don't need to write from prompt, there is still is value over having write operations in the abstraction: users will be able to easily swap a vector DB (during the test phase) in their data update code, just like they'd be able to do it in their prompt integration logic. There may also be other uses inside SK - outside the prompt - where doing updates to the vector database is useful.

Reply:
> This might indeed be tricky given that you have quite a few implementations which you ideally want to keep working, but I'd at least give this a solid try.

This is an amazing blessing, not a curse. It's rare when developing an abstraction that you start out with 10-15 ready-made implementations that can be used immediately to vet the design of that abstraction. Demonstrating that it works with all of these will help prove it out immensely. We should only be so lucky in other areas :)

Reply:
> I'd recommend trying to avoid introducing a **new** abstraction, since that would require reimplementing connectors for all the vector databases, and then maintaining the two implementations for each one etc.

So, if we will try to refactor current `IMemoryStore`, this will definitely require refactoring of all currently existing connectors to support it, while new abstraction can be applied to only a couple of them to see how it goes. 

On the other hand, @stephentoub is right, having 12 already implemented connectors will allow us to find common denominator for all of them and create good abstraction.

> IIRC the abstraction is still considered experimental and therefore can be changed in breaking ways, right?

Yes, we should be able to introduce breaking changes for that, since it's experimental at the moment.